### ISSUE TITLE
How to know if the extension state is remote

### ISSUE THREAD

**Issue by dayo09** [Samsung/ONE-vscode#1209]
## What?

Let's understand and find out the most accurate way to know the extension state

From https://github.com/Samsung/ONE-vscode/pull/1207#discussion_r942275882_

@jyoungyun 
> The `remoteName` is defined by extensions. There are many cases that can be included in this value. If it is remote, it can't use the trash, so it would be better to process it like this.

@dayo09
> I agree :-D But I was not sure if `vscode.env.remoteName` would be undefined when it's run on local.

@dayo09
> Actually I cannot fully understand what the line means. Maybe I should find by `extensionKind`?
> 
> ![image](https://user-images.githubusercontent.com/17171963/183877590-a3bca211-ef2c-4af2-9be6-a34d724c7128.png)
> 
> extensionKind in our package.json is 'workspace'. ![image](https://user-images.githubusercontent.com/17171963/183878543-3cf78836-e805-44d2-a881-8f08c54c608f.png)

@jyoungyun 
> I also think it would be better to use `extensionKind` option. After I defined `extensionKind` as `ui` in our package.json file, I created a vsix package and installed it within the docker extension. Then, the `extensionKind` value was changed to `workspace`. But we need to investigate it more. :)

@dayo09
> Oh I just figured out what it means.
> 
> It concerns the case that our extension is installed in both local and remote. If so, the value will be always 'ssh-remote' even if we open the local workspace. As our extension defines 'extensionKind' as 'workspace', it will run an extension only depending on where the workspace is (remote or local)

@dayo09
> It seems
> 
> (1) Case 1. Local host only remoteName : undefined
> 
> (2) Case 2. Remote only remoteName : ssh-remote (attached-container, for docker) extensionKind: Workspace
> 
> (3) Case 3. Local extension and remote extension both installed Because our exensionKind in 'workspace', it will run dependes on which directory we opens on vscode
> 
> * extensionKind: workspace (when I opened the workspace in remote)
>   
>   * remoteName: ssh-remote
>   * extensionKind: Workspace
> * Local: when I opened the workspace in the local window pc
>   
>   * remoteName: undefined (My expectation was that remoteName is ssh-remote here too, by [[OneExplorer] Fix the bug of delete command in ssh-remote env #1207 (comment)](https://github.com/Samsung/ONE-vscode/pull/1207#discussion_r942283543) 😮‍💨 )
>   * extensionKind: ui

@jyoungyun
> What is the value in this case?
> 
> > (1) Case 1. Local host only
> > remoteName : undefined

@dayo09
> @jyoungyun It was UI. (Well I just uninstalled the remote installed one-vscode and tested in my window pc, it's UI, as same as `Case 3-Local`.)
> 
> I think that `~ the value is defined in all extension hosts (local and remote) in case a remote extension host exists.` is not about Case 3.

**Comment by dayo09** [comment]
Well now it looks like 'extensionKind: ui' means it's running on local PC, and 'extensionKind: workspace' means it's running on remote PC..

**Comment by hyunsik-yoon** [comment]
This link might help to understand `extensionKind`: https://code.visualstudio.com/api/advanced-topics/extension-host

**Comment by jyoungyun** [comment]
Then it would be better to use `extensionKind` variable instead of `remoteName` for distinguishing which environment it's running in. :)

**Comment by dayo09** [comment]
@hyunsik-yoon I went through it but it was still ambiguous for me to understand.. 🤔 I will read it and try to find it out based on those trials.

**Comment by dayo09** [comment]
> This link might help to understand extensionKind: https://code.visualstudio.com/api/advanced-topics/extension-host

The reference from the link says:

extensionKind is a property in the extension manifest, which allows users to specify preferred running location. It can be an array, like ['workspace', 'ui'].

So it's an explanation about `extensionKind in package.json`. It's DIFFERENT FROM `extensionKind in Extension object`.

extensionKind: [ExtensionKind](https://code.visualstudio.com/api/references/vscode-api#ExtensionKind)

> The extension kind describes if an extension runs where the UI runs or if an extension runs where the remote extension host runs. The extension kind is defined in the package.json-file of extensions but can also be refined via the remote.extensionKind-setting. When no remote extension host exists, the value is [UI](https://code.visualstudio.com/api/references/vscode-api#ExtensionKind.UI).

This documentation says Extension<T>.extensionKind will be set depending on package.json or remote.extensionKind-setting. And if there is no remote, it's set as UI. 

To summarize, we can use `Extension<T>.extensionKind` runtime value as the eventual environment setting.

**Comment by hyunsik-yoon** [comment]
> extensionKind: [ExtensionKind](https://code.visualstudio.com/api/references/vscode-api#ExtensionKind)

@dayo09 , the above link does not have the description below.

>> The extension kind describes if an extension runs where the UI runs or if an extension runs where the remote extension host runs. The extension kind is defined in the package.json-file of extensions but can also be refined via the remote.extensionKind-setting. When no remote extension host exists, the value is [UI](https://code.visualstudio.com/api/references/vscode-api#ExtensionKind.UI).

Could you let us know the URL where we can see the above description? Thanks.

**Comment by dayo09** [comment]
@hyunsik-yoon 

https://code.visualstudio.com/api/references/vscode-api#Extension%3CT%3E

![image](https://user-images.githubusercontent.com/17171963/186150384-2d4a7954-0ee9-49e6-8043-e78cde5a5f1b.png)

Oh, I forgot to attach the link. 
This is it. :-D

**Comment by dayo09** [comment]
Now that we know how to distinguish local and remote with `extensionKind`, let's close the issue


https://github.com/Samsung/ONE-vscode/blob/aae9af4463f9b839039e98a5e0c00ef2946351bd/src/extension.ts#L37-L48
